iree/compiler/Dialect/Stream/IR/StreamDialect.cpp

```diff
@@ -57,6 +57,34 @@ struct StreamFolderInterface : public DialectFoldInterface {
   }
 };
 
+// Tries to fold away unrealized_conversion_cast ops if the downstream consumers
+// don't need the extra information. These are inserted during conversion or
+// transforms that may interop with external dialects.
+struct StripResourceConversionCastPattern
+    : public OpRewritePattern<UnrealizedConversionCastOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(UnrealizedConversionCastOp castOp,
+                                PatternRewriter &rewriter) const override {
+    auto result = castOp.getResult(0);
+    if (!result.getType().isa<IREE::Stream::ResourceType>()) return failure();
+    assert(castOp.getNumOperands() == 2 &&
```

Comment:
Can you document in the comments the specific form of unrealized conversion cast you are expecting here and that it is an invariant that this form is the only legal form on these types that can exist.

(sorry: this sketches me out a bit. I convinced myself it is right but not super clear)

Reply:
here: https://github.com/google/iree/pull/7820/commits/8a357d9682f1516a00db856359e26bbc3d6b4812